### PR TITLE
Added 3 boolean operators

### PR DESCRIPTION
**Pull request by Minecraftian14**
Two multioperand operators AND and OR.
One binary operators XOR.

Reference: https://github.com/Judekeyser/calang/issues/6

### CODE REVIEW

File: website/jsCalang.js [hunk 1]
```diff
@@ -94,9 +94,18 @@ class BooleanValue extends TypedValue {
   }
 
   static operatorTable = {
-    "NEGATE": (v, args) => BooleanValue.of(!v)
+    "NEGATE": (v, args) => BooleanValue.of(!v),
+    "AND": (v, args) => BooleanValue.of(v && args.map(BooleanValue.toBoolean).reduce((a,b) => a^b)),
```
Comment (Judekeyser) [inline website/jsCalang.js:98]: This should be XOR

File: calang/calang/Calang.java [hunk 2]
```diff
@@ -46,6 +48,12 @@ int sum(Object[] args) {
         }
         {
             addOperator(BooleanValue.class, "NEGATE", (v, args) -> new BooleanValue(this).with(!v.get()));
+            addOperator(BooleanValue.class, "XOR", (v, args) -> new BooleanValue(this).with(v.get() ^ new BooleanValue(Calang.this).with(args[0]).get()));
```
Comment (Judekeyser) [inline calang/calang/Calang.java:51]: This should have the same unbounded variadic policy as the JS version

**Comment by Judekeyser** [comment]
To me, XOR is associative so could benefit from the same unbounded variadic policy as AND and OR. 

Is it possible to have XAND, IMPLIES too in the PR ?

File: website/jsCalang.js (hunk 1 above)
Comment (Minecraftian14) [reply website/jsCalang.js]: Ah yes, I made a silly mistake (when I was resolving the conflicts).
Thanks a lot 🙏.

File: website/jsCalang.js [hunk 3]
```diff
@@ -94,9 +94,24 @@ class BooleanValue extends TypedValue {
   }
 
   static operatorTable = {
-    "NEGATE": (v, args) => BooleanValue.of(!v)
+    "NEGATE": (v, args) => BooleanValue.of(!v),
+    "XOR": (v, args) => BooleanValue.of(v ^ args.map(BooleanValue.toBoolean).reduce((a,b) => a^b)),
+    "AND": (v, args) => BooleanValue.of(v && args.map(BooleanValue.toBoolean).reduce((a,b) => a&&b)),
+    "OR": (v, args) => BooleanValue.of(v || args.map(BooleanValue.toBoolean).reduce((a,b) => a||b)),
+    "XAND": (v, args) => BooleanValue.of(BooleanValue.xandHelper(v, args.map(BooleanValue.toBoolean).reduce((a,b) => BooleanValue.xandHelper(a,b)))),
```
Comment (Judekeyser) [inline website/jsCalang.js:101]: in this one and the above, we observe a repeating pattern:
```
f(v, args.map(toBoolean).reduce(f)
```
This suggests the introduction of a higher-level function `g` that will match the writing:
```
"AND": (v, args) => g((x,y) => x && y)(v, args)
```
or something similar. Could it be possible to write things that way to lower code-duplication?

File: website/jsCalang.js [hunk 4]
```diff
@@ -94,9 +94,24 @@ class BooleanValue extends TypedValue {
...
+    "IMPLIES": (v, args) => BooleanValue.of(!v || args.map(BooleanValue.toBoolean).reduce((a,b) => !a||b))
   };
 
+  static toBoolean(any) {
```
Comment (Judekeyser) [inline website/jsCalang.js:105]: (In view of the above remark, this method may not be useful. Same remark applies for `of`: after all they seem to serve the same kind of semantic. Can you check they are both still necessary and if so, at least use one in the other?

File: website/jsCalang.js [hunk 5]
```diff
@@ -94,9 +94,24 @@ class BooleanValue extends TypedValue {
...
+    var v = new BooleanValue();
+    v.setValue(any);
+    return v.getValue();
+  }
+
+  static xandHelper(a, b) {
```
Comment (Judekeyser) [inline website/jsCalang.js:111]: (In view of the above remark, this may not be useful anymore)

File: website/jsCalang.js [hunk 6]
```diff
@@ -94,9 +94,24 @@ class BooleanValue extends TypedValue {
   }
 
   static operatorTable = {
-    "NEGATE": (v, args) => BooleanValue.of(!v)
+    "NEGATE": (v, args) => BooleanValue.of(!v),
+    "XOR": (v, args) => BooleanValue.of(v ^ args.map(BooleanValue.toBoolean).reduce((a,b) => a^b)),
+    "AND": (v, args) => BooleanValue.of(v && args.map(BooleanValue.toBoolean).reduce((a,b) => a&&b)),
+    "OR": (v, args) => BooleanValue.of(v || args.map(BooleanValue.toBoolean).reduce((a,b) => a||b)),
+    "XAND": (v, args) => BooleanValue.of(BooleanValue.xandHelper(v, args.map(BooleanValue.toBoolean).reduce((a,b) => BooleanValue.xandHelper(a,b)))),
+    "IMPLIES": (v, args) => BooleanValue.of(!v || args.map(BooleanValue.toBoolean).reduce((a,b) => !a||b))
```
Comment (Judekeyser) [inline website/jsCalang.js:102]: `IMPLIES` operator shouldn't be made variadic I think: it's highly non associative. The risk to be misleading is too high for a first shot in my opinion

Comment (Minecraftian14) [reply website/jsCalang.js:102]: I think I agree that `IMPLIES` shouldn't be made variadic, but leaving it alone too felt just equally disagreeing.

I made a plot to understand what would be a nice way to expect the behavior of "chained implies" and observed the pattern on 6 inputs A,B,C,D,E and F
The following image has two columns in "thick borders"
The first one represents ((((A->B)->C)->D)->E)->F
The second one represents (A->B)&&(B->C)&&(C->D)&&(D->E)&&(E->F)
image

But yeah, I finally concluded, that even if a useful implementation exists for variadic implies, it shouldn't require an "unexpected"/"unconventional" behavior.

So let's drop the variadic implies.
(Btw, did you see anything useful in the plot?)

File: website/jsCalang.js [hunk 7]
```diff
@@ -41,7 +46,7 @@ class TypedValue {
     throw `Unsupported convert {v} from object in {this}`;
   }
 
-  sendMessage(operatorName, args) { return this.operatorTable[operatorName](this.getValue(), args); }
+  sendMessage(operatorName, args) { return this.operatorTable[operatorName](this, args); }
```
Comment (Minecraftian14) [inline website/jsCalang.js:49]: This probably is a "big" change, but please consider this:
The arguments we are providing were "this.getValue()" and a list of TypedValue.
I think, both should be of the _same type_. 

Therefore, I added the removed getValue() from the first argument.

Another motivation to do this was how boolean operators are working.
Given an expression like `A OP B C D`
it was actually interpreted as `A OP (((B OP C) OP C) OP D)`
I believe having `(((A OP B) OP C) OP C) OP D` is better 🤔.

Anyhow, if you are considering these points, let's discuss one more thing.
Should the method provide the TypeValue, or should it map the args with getValue?

**Comment by Judekeyser** [comment]
Well done, thanks @Minecraftian14

**Pull request merged**